### PR TITLE
#164798017 Implement Social Login

### PR DESCRIPTION
**Pull request by vic3king**
#### Description
Currently, users can only signup to authors Haven using the registration forms, it's important we have an extra option to enable a painless signup process by giving users access to signup with their social accounts

#### Type of change
- [x] New feature

#### How Has This Been Tested?
- [x] unit testing
- [x] I also hit the callback endpoint `http://localhost:3000/api/v1/auth/facebook` and received a response from Facebook servers
- [x] I also hit the callback endpoint `http://localhost:3000/api/v1/auth/google` and received a response from google servers
- [x] I also hit the callback endpoint `http://localhost:3000/api/v1/auth/twitter` and received a response from Twitter servers
#### Checklist:
- [x] install passport and related packages
- [x] create controller functions(callback and redirect)
- [x] configure passport services
- [x] write mock test using `nock`
#### PT-ID
[164798017](https://www.pivotaltracker.com/n/projects/2318856/stories/164798016)

#### Screenshots
N/A
####  Background
I used Npm `nock` package to mock a test case for the facebook login.

### CODE REVIEW

File: example.env [hunk 1]
```diff
@@ -4,4 +4,11 @@ DB_DATABASE_DEV=
...
 SECRET=<string>
-
+FACEBOOK_CLIENT_ID=
+FACEBOOK_CLIENT_SECRET=
+TWITTER_APP_ID=
+TWITTER_APP_SECRET=
+GOOGLE_CLIENT_ID=
+GOOGLE_APP_SECRET=
+SESSION_SECRET=
+FRONTEND_URL=
```
Comment (rotimi-babalola) [inline example.env]: Add new line

File: server/app.js [hunk 2]
```diff
@@ -1,14 +1,35 @@
...
 const app = express();
 
 app.use(express.json());
 
 app.use('/api-docs', swaggerUi.serve, swaggerUi.setup(swaggerDocument));
 const port = process.env.PORT || 6000;
+
+passport.use(facebookStrategy);
+app.use(passport.initialize());
+app.use(
```
Comment (rotimi-babalola) [inline server/app.js:22]: What does this do?

File: server/models/user.js [hunk 3]
```diff
@@ -67,6 +67,10 @@ module.exports = (sequelize, DataTypes) => {
     research_field: {
       type: DataTypes.STRING,
     },
+    social: {
+      type: DataTypes.ENUM('facebook', 'twitter', 'google', null),
```
Comment (rotimi-babalola) [inline server/models/user.js]: Why is null one of the ENUM values? @vic3king

Comment (vic3king) [reply server/models/user.js]: > Why is null one of the ENUM values? @vic3king

`null` here is just for when a user signs up using the regular registration process.

Comment (rotimi-babalola) [reply server/models/user.js]: > > Why is null one of the ENUM values? @vic3king
> 
> `null` here is just for when a user signs up using the regular registration process.

Then call it something meaningful like `local`. Null looks weird

File: server/app.js (hunk 2 above)
Comment (vic3king) [reply server/app.js:22]: > What does this do?

`passport.initialize()` is a middleware that has access to the request and response. It initializes passport and It creates a passport instance.

File: server/models/user.js (hunk 3 above)
Comment (vic3king) [reply server/models/user.js]: > > > Why is null one of the ENUM values? @vic3king
> > 
> > 
> > `null` here is just for when a user signs up using the regular registration process.
> 
> Then call it something meaningful like `local`. Null looks weird

Alright

File: server/app.js (hunk 2 above)
Comment (rotimi-babalola) [reply server/app.js:22]: > > What does this do?
> 
> `passport.initialize()` is a middleware that has access to the request and response. It initializes passport and It creates a passport instance.

Ok

**Pull request merged**